errors/error.go

```diff
@@ -4,78 +4,124 @@ import (
 	"encoding/json"
 	"fmt"
 	"net/http"
+
+	"github.com/gojek/fiber/protocol"
+	"google.golang.org/grpc/codes"
 )
 
-// HTTPError is used to capture the error resulting from a HTTP request
-type HTTPError struct {
+// FiberError is used to capture the error resulting from a HTTP request
+type FiberError struct {
 	Code    int    `json:"code"`
 	Message string `json:"error"`
 }
 
-// Error is a getter for the error message in a HTTPError object
-func (err HTTPError) Error() string {
+// Error is a getter for the error message in a FiberError object
+func (err FiberError) Error() string {
 	return err.Message
 }
 
-// ToJSON returns the HTTPError object as a Json encoded byte array
-func (err *HTTPError) ToJSON() ([]byte, error) {
+// ToJSON returns the FiberError object as a Json encoded byte array
+func (err *FiberError) ToJSON() ([]byte, error) {
 	return json.MarshalIndent(err, "", "  ")
 }
 
-// NewHTTPError returns an error of type HTTPError from the input error object.
+// NewFiberError returns an error of type FiberError from the input error object.
 // If the input error is already of the required type, it is returned as is.
 // If not, a generic request failed error is created from the given error.
-func NewHTTPError(err error) *HTTPError {
-	if httpErr, ok := err.(HTTPError); ok {
-		return &httpErr
+func NewFiberError(protocol protocol.Protocol, err error) *FiberError {
+	if fiberError, ok := err.(FiberError); ok {
+		return &fiberError
 	}
-	return ErrRequestFailed(err)
+	return ErrRequestFailed(protocol, err)
 }
 
 var (
-	// ErrRouterStrategyTimeoutExceeded is a HTTPError that's returned when
+	// ErrRouterStrategyTimeoutExceeded is a FiberError that's returned when
 	// the routing strategy fails to respond within given timeout
-	ErrRouterStrategyTimeoutExceeded = &HTTPError{
-		Code:    http.StatusInternalServerError,
-		Message: "fiber: routing strategy failed to respond within given timeout",
+	ErrRouterStrategyTimeoutExceeded = func(protocol protocol.Protocol) *FiberError {
+		statusCode := http.StatusInternalServerError
+		if protocol == "GRPC" {
+			statusCode = int(codes.Internal)
+		}
+		return &FiberError{
+			Code:    statusCode,
+			Message: "fiber: routing strategy failed to respond within given timeout",
+		}
 	}
 
-	// ErrRouterStrategyReturnedEmptyRoutes is a HTTPError that's returned when
+	// ErrRouterStrategyReturnedEmptyRoutes is a FiberError that's returned when
 	// the routing strategy routing strategy returns an empty routes list
-	ErrRouterStrategyReturnedEmptyRoutes = &HTTPError{
-		Code:    http.StatusNotImplemented,
-		Message: "fiber: routing strategy returned empty routes list",
+	ErrRouterStrategyReturnedEmptyRoutes = func(protocol protocol.Protocol) *FiberError {
+		statusCode := http.StatusNotImplemented
+		if protocol == "GRPC" {
+			statusCode = int(codes.Unimplemented)
```

Comment:
Is empty routes a valid configuration? I feel that it should be rejected even when reading the configuration itself. Otherwise we probably should return INVALID_ARGUMENT since it seems configuration issue to me.

Reply:
This isn't for routes, but the strategy which is suppose to return routes. The strategy type is supposed to implement `getRoutesOrder()`. Since strategy can be supplied and register by client, it is not guarantee that it will always return non empty routes.

It is possible that 3 routes are registered and the strategy has its dynamic logic (eg base on time of day) to choose the routes and could end up having no routes returned.

Reply:
I see, I misunderstood it. In that case maybe `NOT_FOUND` make more sense?